pairing/plonk/src/prover.rs

```diff
@@ -0,0 +1,293 @@
+        index: &Index<E>
+    ) -> Result<Self, ProofError>
+    {
+        let n = index.cs.domain.size();
+        if witness.len() != 3*n {return Err(ProofError::WitnessCsInconsistent)}
+
+        let mut oracles = RandomOracles::<E::Fr>::zero();
+        let mut evals = ProofEvaluations::<E::Fr>::zero();
+
+        // the transcript of the random oracle non-interactive argument
+        let mut fq_sponge = EFqSponge::new(index.fq_sponge_params.clone());
+
+        // compute public input polynomial
+        let public = witness[0..index.cs.public].to_vec();
+        let p = -Evaluations::<E::Fr>::from_vec_and_domain(public.clone(), index.cs.domain).interpolate();
+
+        let a = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.l.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+        let b = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.r.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+        let c = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.o.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+
+        // commit to the a, b, c wire values
+        let a_comm = index.urs.get_ref().commit(&a)?;
+        let b_comm = index.urs.get_ref().commit(&b)?;
+        let c_comm = index.urs.get_ref().commit(&c)?;
+
+        // absorb the public input, a, b, c polycommitments into the argument
+        fq_sponge.absorb_fr(&public);
+        fq_sponge.absorb_g(&[a_comm, b_comm, c_comm]);
+
+        // sample beta, gamma oracles
+        oracles.beta = fq_sponge.challenge();
+        oracles.gamma = fq_sponge.challenge();
+
+        // compute permutation polynomial
+
+        let mut z = (0..n).map
+        (
+            |j|
+                (witness[j] + &(index.cs.sigmal[0][j] * &oracles.beta) + &oracles.gamma) *&
+                (witness[j+n] + &(index.cs.sigmal[1][j] * &oracles.beta) + &oracles.gamma) *&
+                (witness[j+2*n] + &(index.cs.sigmal[2][j] * &oracles.beta) + &oracles.gamma)
+        ).collect::<Vec<_>>();
+        
+        algebra::fields::batch_inversion::<E::Fr>(&mut z);
+
+        (0..n).for_each
+        (
+            |j| z[j] *=
+                &((witness[j] + &(index.cs.sid[j] * &oracles.beta) + &oracles.gamma) *&
+                (witness[j+n] + &(index.cs.sid[j] * &oracles.beta * &index.cs.r) + &oracles.gamma) *&
+                (witness[j+2*n] + &(index.cs.sid[j] * &oracles.beta * &index.cs.o) + &oracles.gamma))
+        );
+
+        (1..n).for_each(|i| {let x = z[i-1]; z[i] *= &x});
+        if z.pop().unwrap() != E::Fr::one() {return Err(ProofError::ProofCreation)};
+        z.insert(0, E::Fr::one());
```

Comment:
Not a big deal (maybe it's cleaner this way?) but you could reindex things by 1 so you don't have to insert at the front of the vector (I think that's what's happening here?) which is O(n)

Reply:
I will think of something.